harness/determined/exec/prep_container.py

```diff
@@ -54,22 +59,125 @@ def trial_prep(info: det.ClusterInfo, cert: certs.Cert) -> None:
-
+def do_rendezvous_rm_provided(
+    info: det.ClusterInfo, cert: certs.Cert, resources_id: str
+) -> "det.RendezvousInfo":
     r = request.get(
         info.master_url,
         path=f"/api/v1/allocations/{info.allocation_id}/resources/{resources_id}/rendezvous",
         cert=cert,
     )
-
     jri = r.json()["rendezvousInfo"]
-    rendezvous_info = det.RendezvousInfo(
-        container_addrs=jri["addresses"], container_rank=jri["rank"]
+    addrs, rank = jri["addresses"], jri["rank"]
+    return det.RendezvousInfo(container_addrs=addrs, container_rank=rank)
+
+
+def do_rendezvous_slurm(
+    info: det.ClusterInfo, cert: certs.Cert, resources_id: str
+) -> "det.RendezvousInfo":
+    rank_str = os.environ.get("SLURM_PROCID")
+    assert rank_str, "Unable to complete rendezvous without SLURM_PROCID"
+    rank = int(rank_str)
+
+    num_peers_str = os.environ.get("SLURM_NPROCS")
+    assert num_peers_str, "Unable to complete rendezvous without SLURM_NPROCS"
+    num_peers = int(num_peers_str)
+
+    rendezvous_ip, resolution_error = None, None
+    for rendezvous_iface in rendezvous_ifaces():
+        try:
+            rendezvous_ip = get_ip_from_interface(rendezvous_iface)
+            break
+        except ValueError as e:
+            resolution_error = e
+    if not rendezvous_ip:
+        raise resolution_error or ValueError("unable to resolve rendezvous ip")
+
+    r = request.post(
+        info.master_url,
+        path=f"/api/v1/allocations/{info.allocation_id}/all_gather",
+        cert=cert,
+        json={
+            "request_uuid": uuid.uuid4(),
+            "num_peers": num_peers,
+            "data": {
+                "rendezvous_ip": rendezvous_ip,
+            },
+        },
     )
+    addrs = [d["rendezvous_ip"] for d in r.json()["data"]]
+    return det.RendezvousInfo(container_addrs=addrs, container_rank=rank)
+
+
+def rendezvous_ifaces() -> List[str]:
+    # First case is a manual override. For maximum flexibility, this can be a comma-delimited list.
+    rendezvous_iface = os.environ.get("DET_SLURM_RENDEZVOUS_IFACE")
+
+    # If it doesn't work, fallback to just eth. Rendezvous over eth is fine since horovod will
+    # still use DET_INTER_NODE_NETWORK_INTERFACE for everything important, and SSH over IB mostly
+    # won't work.
+    if not rendezvous_iface:
+        rendezvous_iface = get_eth_interface_name()
+
+    # On systems where there is no eth, DET_INTER_NODE_NETWORK_INTERFACE should work, though.
+    if not rendezvous_iface:
+        rendezvous_iface = os.environ.get("DET_INTER_NODE_NETWORK_INTERFACE")
+
+    # If none of these resolved, we're out of luck.
+    if not rendezvous_iface:
+        raise ValueError("unable to resolve rendezvous iface")
+
+    return rendezvous_iface.split(",")
+
+
+def get_eth_interface_name() -> Optional[str]:
+    net_if_addrs = list(psutil.net_if_addrs())
+    for interface in net_if_addrs:
+        if interface.startswith("eth"):
+            return cast(str, interface)
```

Comment:
can you do:
```python
assert isinstance(interface, str)
return interface
```
instead?

cast() offers no type safety at runtime or lint time... it's really pretty useless.

Reply:
yeah, this was just me saying 'mypy sshhh I know it is a str'